eqllib/analytics/defense-evasion/T1099-timestomp-touch.toml

```
@@ -0,0 +1,20 @@
+[analytic.metadata]
+categories = ["detect"]
+confidence = "high"
+contributors = ["Tony Lambert"]
+created_date = "05/17/2019"
+description = "Identifies use of touch to manipulate timestamps. Adversaries use this technique to hinder filesystem forensic analysis."
+id = "a986fac4-792e-11e9-8f5c-d46d6d62a49e"
+name = "Timestomping via Touch"
+os = ["linux"]
+references = []
+tactics = ["Defense Evasion"]
+tags = ["atomicblue"]
+techniques = ["T1099"]
+updated_date = "05/17/2019"
+
+[analytic]
+query = '''
+process where subtype.create and
+  process_name == "touch" and command_line == "* -acmr *"
```

I don't know much about all of the flags for touch, but I'm wondering if this would be too brittle to match like this. You can use regex character classes, which should help make it more resilient. Here's an example that may need some more tweaking.
```
match(command_line, ?".*? -\w*[acmr].*?")
```

---

Hey @rw-access , most of the opportunistic malware scripts I've come across in the wild use the order `-acmr` but you're absolutely correct that it's likely too brittle. I get absolutely lost on regex, so if you know a way to implement `acmr` in any order that'd work nicely.

---

Yeah regex can definitely get gnarly. I'm thinking this should capture multiple permutations
```suggestion
  process_name == "touch" and
    match(command_line, ?"* -\w*a") and 
    match(command_line, ?"* -\w*c") and 
    match(command_line, ?"* -\w*m") and 
    match(command_line, ?"* -\w*r") and 
```

I think this should work as well, and is more concise but potentially less readable.
```suggestion
  process_name == "touch" and match(command_line, ?"* -(\w*[acmr]){4}")
```

---

Hi,

A few things:
* `-r` requires a parameter so, when joined, will _always_ be the 4th argument
* We should also account for these being split up as `-a -c -m -r` and in any order
* We should also account for the long names
  - `-c` == `--no-create`
  - `-r` == `--reference=`
* We also don't want the regex to spill over into other matching long names which don't apply here
  - _--version_
  - _--time_

With these considerations, the best way to write this would be something similar to (broken out for better readability):

```
process_name == "touch" and 
    (
          // `a` as part of the group or alone
	  match(command_line, ?" -[mc]{0,3}a(\s|$)") and 

	  // `r` has to be last or separate to provide paramter
	  wildcard(command_line, "* -r *", "* --reference=*")
    ) 
    and
    (
          // `m` as part of the group or alone
	  match(command_line, ?" -[ac]{0,3}m(\s|$)") and 

	  // `r` has to be last or separate to provide paramter
	  wildcard(command_line, "* -r *", "* --reference=*")
    ) 
    and
    (
          // `c` as part of the group, alone, or in long form
	  (match(command_line, ?" -[am]{0,3}c(\s|$)") or command_line == "* --no-create *") and 

	  // `r` has to be last or separate to provide paramter
	  wildcard(command_line, "* -r *", "* --reference=*"))
    ) 
    and
    (
          // `r` as part of the group, alone, or in long form
	  match(command_line, ?" -[amc]{0,3}r=\w+") or command_line == "* --reference=*"
    )
```